src/gsi/core_disp_func.F90

```diff
@@ -0,0 +1,1055 @@
+#ifdef FCA_REF_MOD
+module core_disp_func_m
+!$$$ module documentation block
+!           .      .    .                                       .
+! module:   core_disp_func_m
+!  prgmmr: Nehrkorn
+!
+! abstract: Core displacement functionality routines, nonlinear (NLM) version
+!
+! program history log:
+!
+! subroutines included:
+!   sub compute_xy_orig
+!   sub compute_xy_orig_dm
+!   sub exchange_needed
+!   sub exchange_interpolated
+!   sub apply_disp_2d
+!   SUB APPLY_VERT
+!   sub allocate_disp
+!   sub deallocate_disp
+!
+! functions included:
+!   bicub_interp
+!   cubicInterpolate
+!   bilin_interp
+!   init_meta
+!
+! variable definition:
+!
+! attributes:
+!   language: f90
+!   machine:
+!
+!$$$ end documentation block
+  use fp_types_m, only: fp
+  use core_disp_types_m, only: fca_gridded_meta, fca_gridded_disp, &
+       ids,ide,jds,jde,kds,kde, ims,ime,jms,jme,kms,kme, &
+       its,ite,jts,jte,kts,kte, ips,ipe,jps,jpe,kps,kpe, &
+       bilinear, bicubic
+  implicit none
+  private
+  public :: bilin_interp, bicub_interp, cubicInterpolate, compute_xy_orig, apply_disp_2d, apply_vert
+  public :: init_meta, allocate_disp, deallocate_disp
+#ifdef DM_PARALLEL
+  public :: compute_xy_orig_dm
+#endif
+  contains
+#else
+#define TRACE_USE
+#endif
+
+real(fp) function bicub_interp(fin,i,j,dx,dy)
+!$$$  subprogram documentation block
+!                .      .    .                                       .
+! subprogram:    bicub_interp
+!   prgmmr: Nehrkorn
+!
+! abstract:
+!
+! program history log:
+!
+!   input argument list:
+!     fin, dx, dy
+!     i, j
+!
+!   output argument list:
+!
+! attributes:
+!   language: f90
+!   machine:
+!
+!$$$ end documentation block
+  implicit none
+  real(fp), intent(in) :: fin(ims:ime,jms:jme), dx, dy
+  integer, intent(in) :: i, j
+
+  real(fp) :: yy(-1:2)
+  integer :: jj
+
+  do jj=-1,2
+     yy(jj) = cubicInterpolate(fin((i-1):(i+2),j+jj),dx)
+  end do
+  bicub_interp = cubicInterpolate(yy,dy)
+end function bicub_interp
+
+real(fp) function cubicInterpolate(f,x)
+!$$$  subprogram documentation block
+!                .      .    .                                       .
+! subprogram:    cubicInterpolate
+!   prgmmr: Nehrkorn
+!
+! abstract:
+!
+! program history log:
+!
+!   input argument list:
+!     f
+!     x
+!
+!   output argument list:
+!
+! attributes:
+!   language: f90
+!   machine:
+!
+!$$$ end documentation block
+  implicit none
+  real(fp), intent(in) :: f(-1:2) !values at x=-1,0,1,2
+  real(fp), intent(in) :: x ! value in [0,1]
+
+!!$    cubicInterpolate = f(0) + 0.5 * &
+!!$         x*(f(1) - f(-1) + &
+!!$         x*(2.0*f(-1) - 5.0*f(0) + 4.0*f(1) - f(2) + &
+!!$         x*(3.0*(f(0) - f(1)) + f(2) - f(-1))))
```

Comment:
Please remove commented out code unless it will be used in the future.

Reply:
Done.